### PR TITLE
Implement distinct Flush types for Compress::compress vs Decompress::decompress

### PR DESCRIPTION
**Pull request by chrisvittal**
This implements distinct flush types (`FlushCompress` and `FlushDecompress`) for `Compress::compress` and `Decompress::decompress` as in #79. It also removes the old `Flush` type.

To keep the code generic in `read<R: BufRead, D: Ops>` and `Writer<W: Write, D: Ops>`, this adds an internal `Flush` trait for abstracting over the common values of the new flush types. This new `Flush` trait is used as a bound for a new associated type (also called `Flush`) for `Ops`.

### CODE REVIEW

File: src/mem.rs [hunk 1]
```diff
@@ -114,7 +110,60 @@ pub enum Flush {
...
+    /// it should only be used when necessary. This will complete the current
+    /// deflate block and follow it with an empty stored block.
+    Sync = ffi::MZ_SYNC_FLUSH as isize,
+
+    /// Pending input is processed and pending output is flushed.
+    ///
+    /// The return value may indicate that the stream is not yet done and more
+    /// data has yet to be processed.
+    Finish = ffi::MZ_FINISH as isize
+}
```
Comment (alexcrichton) [inline src/mem.rs]: While we're at it, both of these flush enums are intended to be extensible in the sense that client code shouldn't match on them exhaustively. To accomplish this mind adding a `__Nonexhaustive` variant tagged with `#[doc(hidden)]` to both enums? That way it should hopefully send a signal of "we may extend this in the future".

**Comment by alexcrichton** [comment]
👍 Thanks! I'll coordinate this with the 1.0 release so I won't merge this just yet, but rest assured I'll take it from here :)

File: src/mem.rs [hunk 2]
```diff
@@ -91,18 +103,6 @@ pub enum Flush {
...
-    /// current block are held to be written as the next byte after the next
-    /// deflate block is completed.
-    ///
-    /// In this case the decompressor may not be provided enough bits at this
-    /// point in order to complete decompression of the data provided so far to
-    /// the compressor, it may need to wait for the next block to be emitted.
-    /// This is for advanced applications that need to control the emission of
-    /// deflate blocks.
-    Block = ffi::MZ_BLOCK as isize,
-
```
Comment (fstirlitz) [inline src/mem.rs]: Why was this removed? The commit message doesn't explain.

Comment (alexcrichton) [inline src/mem.rs]: @fstirlitz I believe it's because it's not actually used in the backend, but do you have a use case for it?

**Comment by alexcrichton** [comment]
Indeed yeah thanks! I just need to allocate some time to go through the final steps of releasing 1.0 and then publish 1.0. I'll be sure to merge this before that though!

**Comment by alexcrichton** [comment]
Alright I'm doing some prep work so I've merged, thanks so much again @chrisvittal for taking this on!
